h3/src/stream.rs

```diff
@@ -10,27 +10,152 @@ use crate::{
+/// this type makes it possible to prefix wire data with the `StreamType`.
+///
+/// Conveying frames as `Into<WriteBuf>` makes it possible to encode only when generating wire-format
+/// data is necessary (say, in `quic::SendStream::send_data`). It also has a public API ergonomy
+/// advantage: `WriteBuf` doesn't have to appear in public associated types. On the other hand,
+/// QUIC implementers have to call `into()`, which will encode the header in `Self::buf`.
+pub struct WriteBuf<B>
+where
+    B: Buf,
+{
+    buf: [u8; WRITE_BUF_ENCODE_SIZE],
+    len: usize,
+    pos: usize,
+    frame: Option<Frame<B>>,
+}
+
+impl<B> WriteBuf<B>
+where
+    B: Buf,
+{
+    fn encode_stream_type(&mut self, ty: StreamType) {
+        let mut buf_mut = &mut self.buf[self.len..];
+        ty.encode(&mut buf_mut);
+        self.len = WRITE_BUF_ENCODE_SIZE - buf_mut.remaining_mut();
+    }
+
+    fn encode_frame_header(&mut self) {
+        if let Some(frame) = self.frame.as_ref() {
+            let mut buf_mut = &mut self.buf[self.len..];
+            frame.encode(&mut buf_mut);
+            self.len = WRITE_BUF_ENCODE_SIZE - buf_mut.remaining_mut();
+        }
+    }
+}
+
+impl<B> From<StreamType> for WriteBuf<B>
+where
+    B: Buf,
+{
+    fn from(ty: StreamType) -> Self {
+        let mut me = Self {
+            buf: [0; WRITE_BUF_ENCODE_SIZE],
+            len: 0,
+            pos: 0,
+            frame: None,
+        };
+        me.encode_stream_type(ty);
+        me
+    }
+}
+
+impl<B> From<Frame<B>> for WriteBuf<B>
+where
+    B: Buf,
+{
+    fn from(frame: Frame<B>) -> Self {
+        let mut me = Self {
+            buf: [0; WRITE_BUF_ENCODE_SIZE],
+            len: 0,
+            pos: 0,
+            frame: Some(frame),
+        };
+        me.encode_frame_header();
+        me
+    }
+}
+
+impl<B> From<(StreamType, Frame<B>)> for WriteBuf<B>
+where
+    B: Buf,
+{
+    fn from(ty_stream: (StreamType, Frame<B>)) -> Self {
+        let (ty, frame) = ty_stream;
+        let mut me = Self {
+            buf: [0; WRITE_BUF_ENCODE_SIZE],
+            len: 0,
+            pos: 0,
+            frame: Some(frame),
+        };
+        me.encode_stream_type(ty);
+        me.encode_frame_header();
+        me
+    }
+}
+
+impl<B> Buf for WriteBuf<B>
+where
+    B: Buf,
+{
+    fn remaining(&self) -> usize {
+        self.len - self.pos
+            + self
+                .frame
+                .as_ref()
+                .and_then(|f| f.payload())
+                .map_or(0, |x| x.remaining())
+    }
+
+    fn chunk(&self) -> &[u8] {
+        if self.len - self.pos > 0 {
+            &self.buf[self.pos..self.len]
+        } else if let Some(payload) = self.frame.as_ref().and_then(|f| f.payload()) {
+            payload.chunk()
+        } else {
+            &[]
+        }
+    }
+
+    fn advance(&mut self, cnt: usize) {
+        let remaining = self.len - self.pos;
+        if remaining > 0 {
+            assert!(cnt <= remaining);
```

Comment:
I think this means that we can only advance the frame header first, and then a second call is needed to advance the payload. This probably won't happen unless `chunks_vectored` is implemented, but I'd recommend it, assuming the underlying QUIC library can use writev (`send_mmsg`, etc).

Reply:
Goo catch ! I think it was actually a bug. :sweat_smile:. Thanks !

Done in 0a6e000